### PR TITLE
feat(puppeteer-core): Infer element type from complex selector

### PR DESCRIPTION
**Pull request by gomain**
**What kind of change does this PR introduce?**

Better type inference.

**Did you add tests for your changes?**

~Not yet.~ Yes.

**If relevant, did you update the documentation?**

Not yet.

**Summary**

Currently methods that return an element handle, i.e. `.$`, `.waitForSelector` attempt to infer the node element type from the selector string. However, this only works when the selector is an exact match of the element tag, i.e. a selector `"a"` would be inferred as `HTMLAnchorElement` . And not when the selector is complex, i.e. selectors `"a#some-id"`, `div > a`, `a:nth-child(2)` would all fallback on `Element`.

This is due to simply looking up the the selector in `HTMLElementTagNameMap` and `SVGElementTagNameMap` without any attempt to parse the selector string.

This PR is an attempt to do so.

**Does this PR introduce a breaking change?**

This could break existing incorrect assertions using the `as` keyword.

**Other information**

~This PR introduces a dependency on the `type-fest` package.~

This PR is far from complete (no tests, no docs). Put out early for feedback and discussion.

### CODE REVIEW

File: packages/puppeteer-core/src/common/types.ts [hunk 1]
```diff
@@ -17,6 +17,7 @@
 import {JSHandle} from './JSHandle.js';
 import {ElementHandle} from './ElementHandle.js';
 import {LazyArg} from './LazyArg.js';
+import type {LastArrayElement, Split} from 'type-fest';
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/types.ts]: is it possible to implement two helpers that are needed instead of bringing in the entire lib?

Comment (gomain) [reply packages/puppeteer-core/src/common/types.ts]: done.

**Comment by OrKoN** [comment]
@jrandolf could you please review?

**Comment by OrKoN** [comment]
@jrandolf can we have it as an intermediate solution?

**Comment by jrandolf** [comment]
@gomain Could you add test cases for these types in the `test-d` folder?

**Comment by OrKoN** [comment]
@gomain I think tests can be added along existing tests https://github.com/puppeteer/puppeteer/blob/main/test-d/ElementHandle.test-d.ts#L8 E.g., we don't need to use NodeFor directly but can you the public API and check the types.

**Comment by gomain** [comment]
There are a few concerns I think we should discuss.

1. `NodeFor` is public, and therefore deserves to be tested. I'm not too sure as to why it is public though. Suggesting it is not used directly, and therefore not need testing, suggests it should be private.
2. Types influenced by `NodeFor` spans across a large number of api. Picking one or two for the purpose of testing that node type is derived from given selector is at best weak. Covering each and every instance where some selector depicts some type would be superfluous. Wouldn't we rather test `NodeFor` in isolation and take for granted that types built on top of it are correct, given the _on top_ building are trivially simple and/or tested.
3. Implementing tests _along existing tests_ would suffer the limitation of `tsd`
    > The current type testing utility `tsd` is insufficient in asserting a
    > type is strictly of a super type and not one of its sub types.
    > ```ts
    > declare const node: Element;
    > declare const a: HTMLAnchorElement;
    > 
    > expectType<Element>(node); // pass
    > expectType<Element>(a); // also pass
    > ```
4. Unless types are the primary artifact of a library or that .d.ts files were hand written, it does not make sense to test .d.ts files, i.e. after compilation. Once types are generated by a project, it is the description of that project's api. Testing that the type of the api meets some expected type is questioning the spec, not the implementation. This does not dismiss the need for type testing. For the sake of organization, we would want to separate tests - including type tests, from the projects that implement them. For this I propose:
    1. Have a dedicated project `puppeteer-test-types` for type testing. This project does not emit any build.
    2. This project uses `expect-type` for type testing. (Or other compile-time type testing utility)
    3. This project is the last but to `puppeteer-test` in the dependency build chain. If `puppeteer-test-types` fails to build (i.e. typecheck), build fails.
    4. Drop use of `tsd`. Partly because of the above limitation. But also to put type testing in the realm of building.

**Comment by OrKoN** [comment]
@gomain that kind of refactoring is not a priority for us right now. As for this change, I think it'd be sufficient if we add tests like `expectType<ElementHandle<HTMLAnchorElement> | null>(await handle.$('.complex > .selector a'));` (I don't think tsd limitation affects that as we are testing for a subtype).

**Comment by gomain** [comment]
@OrKoN Any comments on points 1. & 2. ?

**Comment by OrKoN** [comment]
@gomain 1) private types are stripped from the resulting .d.ts file so NodeFor needs to be marked public to be available in return types. If we can test it with dts, let's. Otherwise, test-d tests suffice. 2) test-d serves as a e2e test suite to prevent regressions so at least capturing the expectations for some of the methods is useful.

**Comment by OrKoN** [comment]
@gomain what was the problem with using tsd?

```
import {NodeFor} from 'puppeteer';
import {expectType} from 'tsd';

type TestFn = <Selector extends string>(selector: Selector) => NodeFor<Selector>;

const test = (() => {}) as TestFn;

expectType<HTMLAnchorElement>(test('.cls a'));
```

Seems to work fine for me? Also, using Element instead of a more specific type throws an error: `Parameter type Element is declared too wide for argument type HTMLAnchorElement.`

**Comment by OrKoN** [comment]
Sorry if I am misunderstanding the issue but it seems to work?:

```
import {NodeFor} from 'puppeteer';
import {expectType, expectNotType} from 'tsd';

type TestFn = <Selector extends string>(selector: Selector) => NodeFor<Selector>;

const test = (() => {}) as TestFn;

expectType<HTMLAnchorElement>(test('.cls a'));
expectNotType<Element>(test('.cls a'));
expectType<Element>(test('#el'));
```

**Comment by gomain** [comment]
@OrKoN My apologies, false alarm on tsd behaviour. I must have been doing it wrong and jumped to that conclusion too soon.

**Review by OrKoN — APPROVED** [review]

**Comment by OrKoN** [comment]
Many thanks! @gomain
